ts-rs/src/typelist.rs

```diff
@@ -0,0 +1,47 @@
+use std::any::TypeId;
+use std::marker::PhantomData;
+
+use crate::TS;
+
+pub trait TypeVisitor: Sized {
+    fn visit<T: TS + 'static + ?Sized>(&mut self);
+}
+
+
+pub trait TypeList: Copy + Clone {
+    fn push<T: TS + 'static + ?Sized>(self) -> impl TypeList {
+        (self, (PhantomData::<T>,))
+    }
+    fn extend(self, l: impl TypeList) -> impl TypeList {
+        (self, l)
+    }
+
+    fn contains<C: Sized + 'static>(self) -> bool;
+    fn for_each(self, v: &mut impl TypeVisitor);
+}
+
+impl TypeList for () {
+    fn contains<C: Sized>(self) -> bool { false }
+    fn for_each(self, _: &mut impl TypeVisitor) {}
+}
+
+impl<T> TypeList for (PhantomData<T>,) where T: TS + 'static + ?Sized {
+    fn contains<C: Sized + 'static>(self) -> bool {
+        TypeId::of::<C>() == TypeId::of::<T>()
+    }
+
+    fn for_each(self, v: &mut impl TypeVisitor) {
+        v.visit::<T>();
+    }
+}
+
+impl<A, B> TypeList for (A, B) where A: TypeList, B: TypeList {
+    fn contains<C: Sized + 'static>(self) -> bool {
+        self.0.contains::<C>() || self.1.contains::<C>()
+    }
+
+    fn for_each(self, v: &mut impl TypeVisitor) {
+        self.0.for_each(v);
+        self.1.for_each(v);
+    }
+}
```

Comment:
Finally, here is where the induction / recursion is happening:  
For any two `TypeLists` `A` and `B`, `(A, B)` is a `TypeList` as well.

Reply:
That's honestly the most clever use of the type system I've seen! Especially given how Rust usually doesn't like recursive data structures

Reply:
😆 I think I've done much more insane things. For example, you can do real [arithmetic in in the type system alone](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85455e3d09b59b92f30eae8ed1be7b67)